[DBG_X] Compilation failures in L1Trigger/L1TMuonOverlapPhase2

Compilation of L1Trigger/L1TMuonOverlapPhase2 package failed for el8_amd64_gcc12 in CMSSW_14_1_DBG_X_2024-05-09-2300:

```
<gcc>/bin/ld.bfd: tmp/el8_amd64_gcc12/src/L1Trigger/L1TMuonOverlapPhase2/src/L1TriggerL1TMuonOverlapPhase2/ccUO3zNJ.ltrans0.ltrans.o: in function `PtAssignmentNNRegression::PtAssignmentNNRegression(edm::ParameterSet const&, OMTFConfiguration const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)':
  .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<10, 4, 18ul, 3, 13, 16, 4>::input_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<10, 4, 18ul, 3, 13, 16, 4>::lut_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<10, 4, 18ul, 3, 13, 16, 4>::lutSize'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<10, 4, 18ul, 3, 13, 16, 4>::lutOutSum_I'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<10, 4, 18ul, 3, 13, 16, 4>::lutOutSum_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<10, 4, 18ul, 3, 13, 16, 4>::output_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `ap_int_base<4, false>::width'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 1ul, 4, 11, 1, 8>::input_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 1ul, 4, 11, 1, 8>::lut_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 1ul, 4, 11, 1, 8>::lutSize'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 1ul, 4, 11, 1, 8>::lutOutSum_I'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 1ul, 4, 11, 1, 8>::lutOutSum_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 1ul, 4, 11, 1, 8>::output_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `ap_int_base<8, false>::width'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 8ul, 5, 11, 1, 8>::input_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 8ul, 5, 11, 1, 8>::lut_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 8ul, 5, 11, 1, 8>::lutSize'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 8ul, 5, 11, 1, 8>::lutOutSum_I'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 8ul, 5, 11, 1, 8>::lutOutSum_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<5, 11, 8ul, 5, 11, 1, 8>::output_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `ap_int_base<8, false>::width'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::input_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lut_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lutSize'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lutOutSum_I'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lutOutSum_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::output_W'
   <gcc>/bin/ld.bfd: .../CMSSW_14_1_DBG_X_2024-05-09-2300/src/FWCore/MessageLogger/interface/ErrorObj.icc:45: undefined reference to `ap_int_base<5, false>::width'
 collect2: error: ld returned 1 exit status
  gmake: *** [tmp/el8_amd64_gcc12/src/L1Trigger/L1TMuonOverlapPhase2/src/L1TriggerL1TMuonOverlapPhase2/libL1TriggerL1TMuonOverlapPhase2.so] Error 1
``

[full log](https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_14_1_DBG_X_2024-05-09-2300/L1Trigger/L1TMuonOverlapPhase2)

I think `constexpr` should fix the one from cmssw i.e.
```
undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::input_W'
undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lut_W'
undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lutSize'
undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lutOutSum_I'
undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::lutOutSum_W'
undefined reference to `lutNN::LutNeuronLayerFixedPoint<8, 13, 16ul, 5, 11, 9, 5>::output_W'
```

and for  `undefined reference to `ap_int_base<5, false>::width'`, I guess we need to patch https://github.com/Xilinx/HLS_arbitrary_Precision_Types/blob/master/include/ap_int_base.h#L125 to have constexpr ?

--- a/L1Trigger/L1TMuonOverlapPhase2/interface/LutNetworkFixedPointRegression2Outputs.h
+++ b/L1Trigger/L1TMuonOverlapPhase2/interface/LutNetworkFixedPointRegression2Outputs.h
@@ -63,9 +63,9 @@ namespace lutNN {
         LutLayer1;
     LutLayer1 lutLayer1;
 
-    static const unsigned int noHitCntShift = layer1_output_I;  //FIXME should be layer1_output_I ???
+    static constexpr unsigned int noHitCntShift = layer1_output_I;  //FIXME should be layer1_output_I ???
 
-    static const int layer2_input_F = layer1_lut_F;
+    static constexpr int layer2_input_F = layer1_lut_F;
 
     typedef LutNeuronLayerFixedPoint<layer2_input_I,
                                      layer2_input_F,
@@ -77,7 +77,7 @@ namespace lutNN {
         LutLayer2;
     LutLayer2 lutLayer2;
 
-    static const int layer3_input_F = layer2_lut_F;
+    static constexpr int layer3_input_F = layer2_lut_F;
 
     typedef LutNeuronLayerFixedPoint<layer3_input_I,
                                      layer3_input_F,

--- a/L1Trigger/L1TMuonOverlapPhase2/interface/LutNeuronLayerFixedPoint.h
+++ b/L1Trigger/L1TMuonOverlapPhase2/interface/LutNeuronLayerFixedPoint.h
@@ -41,18 +41,18 @@ namespace lutNN {
   template <int input_I, int input_F, size_t inputSize, int lut_I, int lut_F, int neurons, int output_I>
   class LutNeuronLayerFixedPoint {
   public:
-    static const int input_W = input_I + input_F;
-    static const int lut_W = lut_I + lut_F;
+    static constexpr int input_W = input_I + input_F;
+    static constexpr int lut_W = lut_I + lut_F;
 
     //the lut out values sum
     //static const int lutOutSum_I = lut_I + ceil(log2(inputSize)); //MB: ceil(log2(inputSize)) is not constexpr which makes issue for code-checks
-    static const int lutOutSum_I = lut_I + ceillog2(inputSize);
-    static const int lutOutSum_W = lutOutSum_I + lut_F;
+    static constexpr int lutOutSum_I = lut_I + ceillog2(inputSize);
+    static constexpr int lutOutSum_W = lutOutSum_I + lut_F;
 
-    static const int output_W = output_I + lut_F;
+    static constexpr int output_W = output_I + lut_F;
 
     //static_assert( (1<<input_I) <= lutSize);
-    static const size_t lutSize = 1 << input_I;
+    static constexpr size_t lutSize = 1 << input_I;
 
     typedef std::array<ap_ufixed<input_W, input_I, AP_TRN, AP_SAT>, inputSize> inputArrayType;
 

--- a/L1Trigger/L1TMuonOverlapPhase2/src/PtAssignmentNNRegression.cc
+++ b/L1Trigger/L1TMuonOverlapPhase2/src/PtAssignmentNNRegression.cc
@@ -21,35 +21,35 @@
 #include <fstream>
 
 namespace lutNN {
-  static const int input_I = 10;
-  static const int input_F = 4;
-  static const std::size_t networkInputSize = 18;
+  static constexpr int input_I = 10;
+  static constexpr int input_F = 4;
+  static constexpr std::size_t networkInputSize = 18;
 
-  static const int layer1_neurons = 16;
-  static const int layer1_lut_I = 3;
-  static const int layer1_lut_F = 13;
+  static constexpr int layer1_neurons = 16;
+  static constexpr int layer1_lut_I = 3;
+  static constexpr int layer1_lut_F = 13;
 
-  static const int layer1_output_I = 4;
+  static constexpr int layer1_output_I = 4;
   //4 bits are for the count of the noHit layers which goes to the input of the layer2
-  static const int layer2_input_I = 8;
+  static constexpr int layer2_input_I = 8;
 
-  static const int layer2_neurons = 9;
-  static const int layer2_lut_I = 5;
-  static const int layer2_lut_F = 11;
+  static constexpr int layer2_neurons = 9;
+  static constexpr int layer2_lut_I = 5;
+  static constexpr int layer2_lut_F = 11;
 
-  static const int layer3_input_I = 5;
+  static constexpr int layer3_input_I = 5;
 
-  static const int layer3_0_inputCnt = 8;
-  static const int layer3_0_lut_I = 5;
-  static const int layer3_0_lut_F = 11;
-  static const int output0_I = 8;
-  static const int output0_F = 2;
+  static constexpr int layer3_0_inputCnt = 8;
+  static constexpr int layer3_0_lut_I = 5;
+  static constexpr int layer3_0_lut_F = 11;
+  static constexpr int output0_I = 8;
+  static constexpr int output0_F = 2;
 
-  static const int layer3_1_inputCnt = 1;
-  static const int layer3_1_lut_I = 4;
-  static const int layer3_1_lut_F = 11;
-  static const int output1_I = 8;
-  static const int output1_F = 0;  //Does not matter in principle - it is not used
+  static constexpr int layer3_1_inputCnt = 1;
+  static constexpr int layer3_1_lut_I = 4;
+  static constexpr int layer3_1_lut_F = 11;
+  static constexpr int output1_I = 8;
+  static constexpr int output1_F = 0;  //Does not matter in principle - it is not used
 
   typedef LutNetworkFixedPointRegression2Outputs<input_I,
                                                  input_F,

